### PR TITLE
Reject malformed empty boxes and box_index in crop-and-resize kernels

### LINKED ISSUE #123397: tf.raw_ops.CropAndResizeGradImage crashes with rank-1 empty boxes tensor
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf.version.VERSION: 2.21.0-dev0+selfbuilt

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

Python 3.10.12

### Bazel version

bazel-9.1.0

### GCC/compiler version

gcc (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

Calling tf.raw_ops.CropAndResizeGradImage with a rank-1 empty boxes tensor can crash the TensorFlow process instead of raising a clean, catchable TensorFlow exception.

The reproducer below uses only very small tensors:

grads shape:      [0, 1, 1, 1]
boxes shape:      [0]
box_ind shape:    [0]
image_size:       [2, 7, 7, 1]
T:                tf.float32
method:           "bilinear"

Here, boxes is an empty rank-1 tensor:

boxes = tf.constant([], dtype=tf.float32)

For CropAndResizeGradImage, boxes is expected to represent crop boxes, so an empty valid boxes tensor would normally have shape [0, 4]. A rank-1 empty tensor with shape [0] is malformed and should be rejected deterministically.

In my latest TensorFlow build, this input does not fail with a recoverable Python-level exception such as InvalidArgumentError. Instead, the process terminates with a fatal signal.

This can cause a denial-of-service condition when CropAndResizeGradImage receives malformed crop metadata derived from untrusted input. Invalid boxes shapes should be rejected before entering the native image-gradient kernel path.

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

grads = tf.zeros([0, 1, 1, 1], dtype=tf.float32)
boxes = tf.constant([], dtype=tf.float32)
box_ind = tf.constant([], dtype=tf.int32)
image_size = tf.constant([2, 7, 7, 1], dtype=tf.int32)

y = tf.raw_ops.CropAndResizeGradImage(
    grads=grads,
    boxes=boxes,
    box_ind=box_ind,
    image_size=image_size,
    T=tf.float32,
    method="bilinear",
)

tf.reduce_sum(y).numpy()
```

### Relevant log output

```shell
F0000 00:00:1784211510.192957  420821 tensor_shape.cc:45] Check failed: NDIMS == dims() (2 vs. 1) Asking for tensor of 2 dimensions from a tensor of 1 dimensions
*** Check failure stack trace: ***
    @     0x7fe570f936aa  absl::lts_20250814::log_internal::LogMessage::SendToLog()
    @     0x7fe570f935ea  absl::lts_20250814::log_internal::LogMessage::Flush()
    @     0x7fe5708c7e8e  tensorflow::TensorShape::CheckDimsEqual()
    @     0x7fe57060c610  tensorflow::Tensor::tensor<>()
    @     0x7fe5415e634a  tensorflow::CropAndResizeGradImageOp<>::ComputeAsync()::{lambda()#1}::operator()()
    @     0x7fe5415ad036  tensorflow::(anonymous namespace)::RunIfBoxIndexIsValid<>()
    @     0x7fe5415e5961  tensorflow::CropAndResizeGradImageOp<>::ComputeAsync()
    @     0x7fe57055b2d0  tensorflow::AsyncOpKernel::Compute()
    @     0x7fe56f2c42c1  tensorflow::ThreadPoolDevice::Compute()
    @     0x7fe56f24492d  tensorflow::(anonymous namespace)::SingleThreadedExecutorImpl::Run()
    @     0x7fe56f1f564e  tensorflow::FunctionLibraryRuntimeImpl::RunSync()
    @     0x7fe56f20716c  tensorflow::ProcessFunctionLibraryRuntime::RunMultiDeviceSync()
    @     0x7fe56f2103c2  tensorflow::ProcessFunctionLibraryRuntime::RunSync()
    @     0x7fe55a99bb15  tensorflow::KernelAndDeviceFunc::Run()
    @     0x7fe55a917f69  tensorflow::EagerKernelExecute()
    @     0x7fe55a92794d  tensorflow::ExecuteNode::Run()
    @     0x7fe55b2b5abd  tensorflow::EagerExecutor::SyncExecute()
    @     0x7fe55a917343  tensorflow::(anonymous namespace)::EagerLocalExecute()
    @     0x7fe55a912d48  tensorflow::DoEagerExecute()
    @     0x7fe55a919087  tensorflow::EagerExecute()
    @     0x7fe55a20a587  tensorflow::EagerOperation::Execute()
    @     0x7fe55a99643a  tensorflow::CustomDeviceOpHandler::Execute()
    @     0x7fe555043f9f  TFE_Execute
    @     0x7fe51c7d83f2  TFE_Py_FastPathExecute_C()
    @     0x7fe47666699a  pybind11::detail::argument_loader<>::call<>()
    @     0x7fe47666682c  pybind11::cpp_function::initialize<>()::{lambda()#1}::__invoke()
    @     0x7fe476624521  pybind11::cpp_function::dispatcher()
    @     0x55fa03107962  (unknown)
Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by vishwakt**
## Summary

Fixes #123397.

`tf.raw_ops.CropAndResizeGradImage` with a rank-1 empty `boxes` tensor (shape `[0]` instead of `[0, 4]`) aborts the whole process with a fatal CHECK failure instead of raising a catchable error:

```
F ... tensor_shape.cc:47] Check failed: NDIMS == dims() (2 vs. 1) Asking for tensor of 2 dimensions from a tensor of 1 dimensions
```

## Root cause

`ParseAndCheckBoxSizes` returns early with `num_boxes = 0` when both `boxes` and `box_index` are empty, skipping all rank validation. The malformed tensor then reaches `boxes.tensor<float, 2>()` in the compute callback, which CHECK-fails on the rank mismatch and calls `abort()`.

I reproduced this on TF 2.21.0 and found the same defect is reachable through two more vectors sharing the same helper:

- `CropAndResizeGradBoxes` with rank-1 empty `boxes`: same fatal crash.
- `CropAndResizeGradImage` with rank-2 empty `box_ind` (shape `[0, 0]`): fatal crash at `box_index.tensor<int32_t, 1>()`.

The forward `CropAndResize` op already fails cleanly.

## Fix

Validate the ranks of `boxes` and `box_index` before the empty early return in `ParseAndCheckBoxSizes`. The element-count checks (4 columns, matching row counts) stay after the early return, so well-formed empty inputs (`boxes` of shape `[0, 4]`, `box_index` of shape `[0]`) keep working; I verified that path still returns a correctly shaped output.

Adds a regression test covering all three crash vectors plus the well-formed empty case.

Credit to @yyds1233 for the report and the clean reproducer.

### CODE REVIEW

File: tensorflow/core/kernels/image/crop_and_resize_op.cc [hunk 1]
```diff
@@ -55,24 +55,25 @@ using Callback = std::function<void()>;
...
   }
-  *num_boxes = boxes.dim_size(0);
-  if (boxes.dim_size(1) != 4) {
-    return absl::InvalidArgumentError("boxes must have 4 columns");
-  }
-  // The shape of 'box_index' is [num_boxes].
   if (box_index.dims() != 1) {
     return absl::InvalidArgumentError(
         absl::StrCat("box_index must be 1-D", box_index.shape().DebugString()));
   }
```
[the remark this answers is not in the document]
Comment (vishwakt) [reply tensorflow/core/kernels/image/crop_and_resize_op.cc:68]: Added separator

**Review by dmiltr3 — APPROVED** [review]
### Review Details:
1. **Input Validation**: The change correctly moves the rank validation of `boxes` and `box_index` before the early return path for empty inputs. This prevents the kernel from attempting to access rank-1 tensors as rank-2 (or vice versa), which was triggering the fatal `CHECK` failures.
2. **Backward Compatibility**: Valid empty inputs (e.g., `boxes` of shape `[0, 4]` and `box_index` of shape `[0]`) continue to be handled correctly, as the early return path is preserved and triggers after rank validation.
3. **Coverage**: The proposed regression tests effectively cover the identified crash vectors.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi @vishwakt,

Thank you for the pull request and for addressing the empty tensor rank validation in `crop_and_resize`!

The Cloud Fortify Code Security Analyzer checks passed cleanly with **0 findings**. However, the internal TAP presubmit tests failed due to a strict dependencies violation in the test suite:

### 1. Failure Details
- **File:** `tensorflow/python/ops/image_grad_test_base.py`
- **Error:**
  ```text
  ERROR: Strict deps violations: //tensorflow/python/ops:image_grad_test_base
    In tensorflow/python/ops/image_grad_test_base.py, no direct deps found for imports:
      line 27: from tensorflow.python.ops import array_ops: Module not provided by a direct dep
  ```
- **Cause:** `array_ops` was imported in `image_grad_test_base.py`, but `:array_ops` is not declared in the `deps` of `//tensorflow/python/ops:image_grad_test_base` in `tensorflow/python/ops/BUILD`.

---

### 2. How to Fix

Please apply either of the following solutions:

#### Option A (Recommended): Avoid importing `array_ops`
`image_grad_test_base.py` already imports `constant_op` and `numpy as np` (both in `deps`). You can construct the zero tensors using `constant_op.constant(np.zeros(...))` and remove the `array_ops` import:

```python
# 1. Remove this import from image_grad_test_base.py:
# from tensorflow.python.ops import array_ops

# 2. In testMalformedEmptyBoxesRaisesError, construct tensors via constant_op:
  def testMalformedEmptyBoxesRaisesError(self):
    grads = constant_op.constant(np.zeros([0, 1, 1, 1], dtype=np.float32))
    image_size = constant_op.constant([2, 7, 7, 1], dtype=dtypes.int32)
    valid_boxes = constant_op.constant(np.zeros([0, 4], dtype=np.float32))
    valid_box_ind = constant_op.constant(np.zeros([0], dtype=np.int32))
    with self.assertRaisesRegex(
        (errors_impl.InvalidArgumentError, ValueError), 'boxes must be 2-D'
    ):
      self.evaluate(
          gen_image_ops.crop_and_resize_grad_image(
              grads=grads,
              boxes=constant_op.constant(np.zeros([0], dtype=np.float32)),
              box_ind=valid_box_ind,
              image_size=image_size,
              T=dtypes.float32,
          )
      )
    with self.assertRaisesRegex(
        (errors_impl.InvalidArgumentError, ValueError), 'box_index must be 1-D'
    ):
      self.evaluate(
          gen_image_ops.crop_and_resize_grad_image(
              grads=grads,
              boxes=valid_boxes,
              box_ind=constant_op.constant(np.zeros([0, 0], dtype=np.int32)),
              image_size=image_size,
              T=dtypes.float32,
          )
      )
    with self.assertRaisesRegex(
        (errors_impl.InvalidArgumentError, ValueError), 'boxes must be 2-D'
    ):
      self.evaluate(
          gen_image_ops.crop_and_resize_grad_boxes(
              grads=grads,
              image=constant_op.constant(np.zeros([2, 7, 7, 1], dtype=np.float32)),
              boxes=constant_op.constant(np.zeros([0], dtype=np.float32)),
              box_ind=valid_box_ind,
          )
      )
    # Well-formed empty boxes must keep working.
    output = self.evaluate(
        gen_image_ops.crop_and_resize_grad_image(
            grads=grads,
            boxes=valid_boxes,
            box_ind=valid_box_ind,
            image_size=image_size,
            T=dtypes.float32,
        )
    )
    self.assertEqual((2, 7, 7, 1), output.shape)
```

#### Option B: Add `:array_ops` to `tensorflow/python/ops/BUILD`
If you prefer keeping `from tensorflow.python.ops import array_ops`, add `":array_ops"` to `deps` in `tensorflow/python/ops/BUILD`:
```python
py_library(
    name = "image_grad_test_base",
    srcs = ["image_grad_test_base.py"],
    strict_deps = True,
    deps = [
        ":array_ops",  # <-- Add this dependency
        ":array_ops_stack",
        ":gradient_checker_v2",
        ":image_ops",
        ":image_ops_gen",
        ":math_ops",
        "//third_party/py/absl/testing:parameterized",
        "//third_party/py/numpy",
        "//third_party/tensorflow/python/eager:backprop",
        "//third_party/tensorflow/python/eager:context",
        "//third_party/tensorflow/python/framework:config",
        "//third_party/tensorflow/python/framework:constant_op",
        "//third_party/tensorflow/python/framework:dtypes",
        "//third_party/tensorflow/python/framework:errors",
        "//third_party/tensorflow/python/framework:test_lib",
        "//third_party/tensorflow/python/platform:client_testlib",
    ],
)
```

**Review by dmiltr3 — APPROVED** [review]
### Review Details

1. **Input Validation**: The change correctly validates the rank of `boxes` and `box_index` before early return on empty tensors, preventing fatal downstream CHECK failures when malformed empty tensors are passed.
2. **Dependencies & Tests**: Added `:array_ops` dependency in `tensorflow/python/ops/BUILD` for `image_grad_test_base` to satisfy strict dependencies. Regression tests in `image_grad_test_base.py` verify both error handling for invalid ranks and preserved functionality for valid empty inputs.
3. **Tests Status**: All Linux CPU, Linux CUDA, Linux CUDA 13, ARM64, and PyLint checks pass cleanly.
